File: Application/Sources/Helpers/Categories/SRGProgramComposition+PlaySRG.m
//
//  Copyright (c) SRG SSR. All rights reserved.
//
//  License information is available from the LICENSE file.
//

#import "SRGProgramComposition+PlaySRG.h"

#import "PlaySRG-Swift.h"

@implementation SRGProgramComposition (PlaySRG)

- (SRGProgram *)play_programAtDate:(NSDate *)date
{
    NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(SRGProgram * _Nullable program, NSDictionary<NSString *,id> * _Nullable bindings) {
        return [program play_containsDate:date];
    }];
    return [self.programs filteredArrayUsingPredicate:predicate].firstObject;
}

- (NSArray<SRGProgram *> *)play_programsFromDate:(NSDate *)fromDate toDate:(NSDate *)toDate withMediaURNs:(NSArray<NSString *> *)mediaURNs
{
    NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(SRGProgram * _Nullable program, NSDictionary<NSString *,id> * _Nullable bindings) {
        if (fromDate && [program.startDate compare:fromDate] == NSOrderedAscending) {
            return NO;
        }
        
        if (toDate && [toDate compare:program.startDate] == NSOrderedAscending) {
            return NO;
        }
        
        if (mediaURNs && ! [mediaURNs containsObject:program.mediaURN]) {
            return NO;
        }
        
        return YES;
    }];
    
    return [self.programs filteredArrayUsingPredicate:predicate];
}

@end

File: Application/Sources/Helpers/Extensions/SRGProgramComposition+PlaySRG.swift
//
//  Copyright (c) SRG SSR. All rights reserved.
//
//  License information is available from the LICENSE file.
//

import SRGDataProviderModel

extension SRGProgramComposition {
    /**
     *  Return the program at the specified date, if any.
     */
    @objc func play_program(at date: Date) -> SRGProgram? {
        programs?.first(where: { $0.play_containsDate(date) })
    }

    /**
     *  Returns only programs matching in a given date range. The range can be open or possibly half-open. If media URNs
     *  are provided, only matching programs will be returned.
     */
    @objc func play_programs(from fromDate: Date?, to toDate: Date?, withMediaURNs mediaURNs: [String]?) -> [SRGProgram] {
        programs?.filter { program in
            if let fromDate, program.startDate < fromDate {
                return false
            }

            if let toDate, toDate < program.startDate {
                return false
            }

            if let mediaURNs {
                guard let mediaUrn = program.mediaURN else { return false }
                return mediaURNs.contains(mediaUrn)
            }

            return true
        } ?? []
    }
}
